### DISCUSSION TITLE
[STUDIO] ACTEUR Liste de reflets

### DISCUSSION THREAD

**Discussion by WITNicolas** [witsa/synapps#566] [Q&A] [answered]
> Lien de la page : `concepts/actor-types/redy-reflects-list.md`.
(Retour Certifié)
L'acteur liste de reflets affiche les reflets dans un ordre qui n'est pas maitrisé.
Si par exemple, la liste affichée dépend d'un ensemble, il n'y a pas de corrélation entre l'ordre dans l'ensemble et l'ordre affiché.
Comment sont classés les reflets dans la liste ?

Je ne vois pas de lien entre les label, les ID, et l'ordre de la liste.

Il serait interessant que l'ordre soit respecté.
Au niveau de l'exploitation il y a logique d'avoir par exemple d'abord les Marche/Arret, puis les consignes, puis les commandes, etc...etc... et pas tout mélangé.

**Comment by DavMar05** [comment]
En effet pour l'affichage d'un ensemble il serait quand même super bien que studio garde l'ordre du SET et non pas l'ordre de création. Si vous pouvez me trouver une solution car j'ai pas mal de réglages dans les ensembles à mettre en ordre pour le client final. 
merci

**Comment by manufitoussiwit** [comment]
Effectivement, il faudrait pouvoir ordonner par un champ en particulier ou par l'ordre dans un ensemble. Je vais prévenir l'équipe de développement du REDY pour qu'elle étudie la possibilité d'intégrer ça dans la prochaine version de REDY.

**Comment by manufitoussiwit** [comment]
En attendant, je suis en train d'étudier un *petit* script de contournement à ajouter sur l'acteur liste de reflet pour qu'il ordonne les reflets comme ils sont définis dans l'ensemble choisi. 
C'est prometteur 👍 
Je le partagerai demain matin s'il est bien efficace.

**Comment by WITNicolas** [comment]
Haaa ça c'est top s'il y a une petite solution SortBy(asc)  :smiley:

**Comment by DavMar05** [comment]
👍  super

**Comment by manufitoussiwit** [comment, marked as answer]
Voici le contournement  : *(à copier/coller dans l'arborescence d'une scène)*
```text
SYNAPPS-STUDIO-ACTOR|{"type":"redy/reflect-list","key":"reflect-list1","properties":{"pageSize":2000},"additionalDefs":{"sortBy":{"type":"select","label":"Ordonner par","options":"[\n  {\n    \"value\": \"none\",\n    \"text\": \"Rien\"\n  },\n  {\n    \"value\": \"name\",\n    \"text\": \"Nom\"\n  },\n  {\n    \"value\": \"collection\",\n    \"text\": \"Ensemble\"\n  },\n  {\n    \"value\": \"stateText\",\n    \"text\": \"Etat\"\n  },\n  {\n    \"value\": \"type\",\n    \"text\": \"Type\"\n  },\n  {\n    \"value\": \"path\",\n    \"text\": \"Chemin\"\n  },\n  {\n    \"value\": \"allowCommand\",\n    \"text\": \"Commandable?\"\n  },\n  {\n    \"value\": \"hasPendingEvent\",\n    \"text\": \"En cours?\"\n  },\n  {\n    \"value\": \"lastRefreshDate\",\n    \"text\": \"Dernier rafraichissement\"\n  },\n  {\n    \"value\": \"valueDate\",\n    \"text\": \"Dernier changement\"\n  }\n]"},"sortDir":{"type":"select","label":"Direction Ordre","options":"[\n  {\n    \"value\": \"up\",\n    \"text\": \"Montant\"\n  },\n  {\n    \"value\": \"down\",\n    \"text\": \"Descendant\"\n  }\n]"}},"additionals":{"sortBy":"none","sortDir":"up"},"events":{"onInit":["","const loadReflects = this.loadReflects;","const loadReflectsWithSort = async () => {","  const regularResult = await loadReflects.bind(this)();","  switch (this.additionals.sortBy) {","    case 'collection':","      const redy = context.dataStores.redy;","      const collectionLabel = this.properties.collection;","      if (!collectionLabel) return regularResult;","      const collection = await redy.loadWosVariable(`:easy.Set.${collectionLabel}.List`, { withBlob: true });","      if (!collectionLabel) return regularResult;","      const paths = atob(collection.value).split('\\r').filter(p => !context.utils.isEmpty(p));","      const variables = [];","      const getOrLoadWosVariable = async (path) => {","        let v = await redy.getWosVariable(path);","        if (!v) {","          v = await redy.loadWosVariable(path)","        }","","        return v;","      }","      for (const path of paths) {","        variables.push(await getOrLoadWosVariable(path));","      }","","      const ids = variables.map(v => v.redyID);","","      let reflects = ids.map(id => regularResult.rows.find(r => r.redyID === id)).filter(r => r);","      if (this.additionals.sortDir === 'down') {","        reflects.reverse();","      }","","      return { count: regularResult.count, rows: reflects };","","    case 'name':","    case 'stateText':","    case 'path':","    case 'type':","    case 'allowCommand':","    case 'hasPendingEvent':","    case 'valueDate':","    case 'lastRefreshDate':","      const key = this.additionals.sortBy;","      regularResult.rows = regularResult.rows.sort((r1, r2) => String(r1[key] || '').localeCompare(String(r2[key] || '')));","      if (this.additionals.sortDir === 'down') {","        regularResult.rows.reverse();","      }","      return regularResult;","","    default:","      return regularResult;","  }","}","","this.loadReflects = loadReflectsWithSort;"],"onPropertyChanged":["if (['additionals.sortBy', 'additionals.sortDir'].includes(context.propertyPath)) {","  this.refresh();","}"]}}
```

- J'ai rajouté deux additionnelles qui permettent de choisir par quel champ on veut ordonner et dans quelle direction.

![image](https://user-images.githubusercontent.com/35595723/180189590-d956039c-feeb-4db1-8db5-4153741ac18d.png)



![image](https://user-images.githubusercontent.com/35595723/180189517-e67d9a61-9f54-42bf-9f3d-e9fd831c7434.png)


- Pour ordonner comme défini dans un ensemble, il faut bien entendu filtrer au préalable par un ensemble.
- La limite du contournement c'est que tout est fait par script coté client, donc il faut prendre tout les reflets. Prévoir une taille de pagination assez grande (là j'ai mis 2000)

### Démo :
![sortBy](https://user-images.githubusercontent.com/35595723/180188487-025ba46f-2710-41bf-a276-70adde7f1ca3.gif)

## Correction du 16/07/2023

**Reply by manufitoussiwit** [reply]
corrigé grace à @Raphwit

**Comment by WITNicolas** [comment]
Excellent ! Bravo !

**Comment by WIToliv** [comment]
Super ça m'aide, une intégration direct dans la liste de reflet sur une prochaine mise à jour serait top.
merci et bravo

**Comment by Raphwit** [comment]
Bonjour, suites à quelques changements dans Synapps, l'acteur ne fonctionnait plus, voici une version fonctionnelle mise à jour !

```
SYNAPPS-STUDIO-ACTOR|{"type":"redy/reflect-list","key":"reflect-list1","properties":{"pageSize":2000},"additionalDefs":{"sortBy":{"type":"select","label":"Ordonner par","options":"[\n  {\n    \"value\": \"none\",\n    \"text\": \"Rien\"\n  },\n  {\n    \"value\": \"name\",\n    \"text\": \"Nom\"\n  },\n  {\n    \"value\": \"collection\",\n    \"text\": \"Ensemble\"\n  },\n  {\n    \"value\": \"stateText\",\n    \"text\": \"Etat\"\n  },\n  {\n    \"value\": \"type\",\n    \"text\": \"Type\"\n  },\n  {\n    \"value\": \"path\",\n    \"text\": \"Chemin\"\n  },\n  {\n    \"value\": \"allowCommand\",\n    \"text\": \"Commandable?\"\n  },\n  {\n    \"value\": \"hasPendingEvent\",\n    \"text\": \"En cours?\"\n  },\n  {\n    \"value\": \"lastRefreshDate\",\n    \"text\": \"Dernier rafraichissement\"\n  },\n  {\n    \"value\": \"valueDate\",\n    \"text\": \"Dernier changement\"\n  }\n]"},"sortDir":{"type":"select","label":"Direction Ordre","options":"[\n  {\n    \"value\": \"up\",\n    \"text\": \"Montant\"\n  },\n  {\n    \"value\": \"down\",\n    \"text\": \"Descendant\"\n  }\n]"}},"additionals":{"sortBy":"none","sortDir":"up"},"events":{"onInit":["","const loadReflects = this.loadReflects;","const loadReflectsWithSort = async () => {","  const regularResult = await loadReflects.bind(this)();","  switch (this.additionals.sortBy) {","    case 'collection':","      const redy = context.dataStores.redy;","      const collectionLabel = this.properties.collection;","      if (!collectionLabel) return regularResult;","      const collection = await redy.loadWosVariable(`:easy.Set.${collectionLabel}.List`, { withBlob: true });","      if (!collectionLabel) return regularResult;","      const paths = atob(collection.value).split('\\r').filter(p => !context.utils.isEmpty(p));","      const variables = [];","      const getOrLoadWosVariable = async (path) => {","        let v = await redy.getWosVariable(path);","        if (!v) {","          v = await redy.loadWosVariable(path)","        }","","        return v;","      }","      for (const path of paths) {","        variables.push(await getOrLoadWosVariable(path));","      }","","      const ids = variables.map(v => v.redyID);","","      let reflects = ids.map(id => regularResult.rows.find(r => r.redyID === id)).filter(r => r);","      if (this.additionals.sortDir === 'down') {","        reflects.reverse();","      }","","      return { count: regularResult.count, rows: reflects };","","    case 'name':","    case 'stateText':","    case 'path':","    case 'type':","    case 'allowCommand':","    case 'hasPendingEvent':","    case 'valueDate':","    case 'lastRefreshDate':","      const key = this.additionals.sortBy;","      regularResult.rows = regularResult.rows.sort((r1, r2) => String(r1[key] || '').localeCompare(String(r2[key] || '')));","      if (this.additionals.sortDir === 'down') {","        regularResult.rows.reverse();","      }","      return regularResult;","","    default:","      return regularResult;","  }","}","","this.loadReflects = loadReflectsWithSort;"],"onPropertyChanged":["if (['additionals.sortBy', 'additionals.sortDir'].includes(context.propertyPath)) {","  this.refresh();","}"]}}
```

**Reply by manufitoussiwit** [reply]
Merci @Raphwit ! Je met à jour la réponse avec.

**Comment by edouar1** [comment]
Bonjour, 
pourquoi, on peut pas commander depuis la liste des reflet?
Cordialement,

**Reply by WITNicolas** [reply]
Bonjour,
Parce que cet acteur n'est pas fait pour cela, tout simplement.
La liste de reflet liste simplement les reflets.
Pour avoir la commande il faut utiliser le détail de reflet et liaisonner cet acteur à la selection faite dans la liste.

Vous avez la possibilité également de réaliser un composite avec le mécanisme ci-dessus pour créer vous -même votre propre principe de fonctionnement

Je vous invite à suivre https://github.com/witsa/synapps/discussions/725 qui peut être aussi une solution qui conviendrait à ce que vous souhaitez faire.

Dans le principe, tous les acteurs ont des comportements de base que vous pouvez customiser grace au principe des composites pour créer/imaginer toutes interactions entre les acteurs afin de créer justement un comportement plus "complexe". Et pouvoir par la suite réutiliser cela dans tous vos projets.

Cordialement,

**Comment by edouar1** [comment]
parfait, merci pour te retour.
C'est quoi la déférence entre liste état et liste reflet svp?
cordialement,

**Reply by WITNicolas** [reply]
La liste d'Etat est faite pour afficher les Etats des ressources (comme pour les Synoptiques)
La liste de Reflets est faite pour afficher les Reflets qui sont activés dans le REDY (au niveau de chaque ressource) reflet Etat/analogique/digital commandable etc...

**Comment by edouar1** [comment]
Ah oui c'est claire pour moi. 
merci beaucoup.
Cordialement,
